### ISSUE TITLE
Things Regarding Spam

### ISSUE THREAD

**Issue by theabbie** [echojs/echojs#12]
EchoJS used to be a great platform when it was barely popular, now that more people know this community, it still relies on ethics to prevent spamming.

Though all communities can be spammed with some effort, what makes EchoJS worse is lack of proper moderation, and due to that, it is highly prone to spam.

Apart from moderation, There are many factors which make spamming very easy on EchoJS.

* No Email Authentication is a terrible idea, it makes account creation very easy.
* No option to delete your posts and accounts.
* No CAPTCHA, atleast for account creation.
* No "Report Spam" option (downvote is not same as Report spam)

Only Spam protection mechanism is rate limiting account creation by IP Address, That would have been effective if accompanied by a CAPTCHA, but, without that, it's useless.

This is the code snippet for creating a new account and upvoting a post,

```js
var axios = require("axios");

var data = await axios({
  url: 'https://www.echojs.com/api/create_account',
  method: 'POST',
  data: 'username='+username+'&password='+password,
  withCredentials: true
});

console.log(data.data);

await axios({
  url: 'https://www.echojs.com/api/votenews',
  method: 'POST',
  data: 'vote_type=up&apisecret='+data.data.apisecret+'&news_id='+post_id,
  headers: {
    cookie: 'auth='+data.data.auth
  }
});
```

Though limited by IP address, all you have to do is run this code on devices with different IP addresses.

There are lots of online code compilers which run the code on server, new IP address everytime, making the limit useless.

Had this been an Email Authentication and CAPTCHA, it would not be possible to do this through a script.

The decision is up to you, whether you want to run such a community, or, shut it down, atleast temporarily, fix these problems and run it again.

Saying that it's the duty of users to not spam is just transferring the burden. If you refuse to take action now, it's just bad for all users, People can manipulate votes to favour bad posts without ever getting caught.

Thank You For Reading.

**Comment by Inlustra** [comment]
Have you ever heard of responsible disclosure? Have a [read](https://en.wikipedia.org/wiki/Responsible_disclosure).

You could have easily disclosed the vulnerabilities to the site owners. Instead you've been publicly spamming a free, community maintained website, impersonating people (Have another [read](https://www.michalsons.com/blog/liability-impersonating-someone-online/9696)), and yet some how you still believe you're in the right?

You're absolutely correct about one thing, it should be harder to create an account and post. Once that work is complete, you should be disallowed from posting or voting ever again.

**Comment by theabbie** [comment]
@Inlustra I didn't say I did any right, that been said, yes, it's great if the members are ethical, In an ideal world, a community could even run without authentication, people following rules and being ethical. But, Members being ethical is never an excuse for not preventing Spam, An Ideal example is Hacker News, They are doing a very good job while being very similar to EchoJS. If Spam prevention is not the #1 priority while designing a community, that community doesn't deserve to exist. Ability to Spam is a failure on the part of the community, And, EchoJS is clearly a failure.

I was very serious about shutting down EchoJS as these are not some minor issues. EchoJS is fundamentally flawed in it's design.

TLDR Don't blame incompetence on ethics, EchoJS is a failure.

**Comment by vladimyr** [comment]
> You could have easily disclosed the vulnerabilities to the site owners. Instead you've been publicly spamming a free, community maintained website, impersonating people (Have another read), and yet some how you still believe you're in the right?

Or they could address them by proposing PRs with required changes.

> If Spam prevention is not the #1 priority while designing a community, that community **doesn't deserve to exist**. Ability to Spam is a failure on the part of the community, And, **EchoJS is clearly a failure.**

> I was very serious about shutting down EchoJS as these are not some minor issues. EchoJS is fundamentally flawed in it's design.

> TLDR Don't blame incompetence on ethics, **EchoJS is a failure.**

Because you obviously have the absolute moral strength and authority to make those verdicts I'll follow your brilliant example and thereby I _proclaim_:
Thou who are able to code but instead of coding they use harassing as a form of bringing _justice_ to the Internet/s doesn't deserve to have a voice on Github.

**PS** I'm _very serious_ about it, would you please humor me and become a @ghost? 😉

**Comment by theabbie** [comment]
@vladimyr
> Or they could address them by proposing PRs with required changes.

You mean adding Captcha, using Email authentication, And many other features? These are not the changes you could just add by adding some code. It will require a complete redesign. Similar to just Creating a new Community.

Being open-source does not automatically excuse it from not improving. Though anyone can contribute to it, they have the right to demand features. The creators are still responsible for maintaining it, abandoned projects should be marked as archived.

Stop defending this, This Community isn't designed to persist.

**Comment by vladimyr** [comment]
> You mean adding Captcha, using Email authentication, And many other features? These are not the changes you could just add by adding some code. It will require a complete redesign. Similar to just Creating a new Community.

Nobody is preventing you from doing that complete redesign or creating the new best thing since sliced bread. Do you have any troubles with the stuck fork button? Should we contact Github support?

> they have the right to demand features. The creators are still responsible for maintaining it, abandoned projects should be marked as archived.

I'm sure you got that from reading the LICENSE. I was never good at legalese, thank you for enlightenment. 😉

> Stop defending this, This Community isn't designed to persist.

You are right, after years of reading and learning it is time to stop. As soon as the night falls I'll lit my torch, pick my pitchfork and start attacking it. Thank you for pointing me in the right direction.

**Comment by theabbie** [comment]
@vladimyr If at all I ever design a community, I will look down to EchoJS to learn what not to do, Only thing EchoJS will be famous for is being bad. Anyways, Have a nice day.

**Comment by tracker1** [comment]
While there are additional features, such as some of the suggestions that would be nice to have and/or value add.  It's ironic and almost gaslighting considering the source of the issue.

**Comment by biziclop** [comment]
This dude is funny (not):  
https://opensource.stackexchange.com/questions/11240/is-it-wrong-to-demand-features-in-open-source-projects

**Comment by vladimyr** [comment]
> Whenever I demand some feature or complain about some vulnerability, small or big, there is always someone who says "Instead of reporting/demanding this, why not just create a PR with the demands fulfilled".

**Whenever I do exploit some vulnerability, small or big, in a malicious and harmful way and complain about it afterward, there is always...**

There, I fixed it up for you.

I'm sure forgetting to mention that you filled EchoJS frontpage with a ton of disgusting crap and even tried to impersonate other users was completely unintentional. I would also bet that commenters on SE being aware of that _little detail_ wouldn't make that discussion anything less pleasant/constructive.

**Comment by theabbie** [comment]
@vladimyr I am pretty sure that even without all that exploitation, had I informed you all about this, no one could have cared any less about this issue. Maybe it would have been more polite, but, nothing more.

Now that everyone has got a taste of Spam, they know the problem. Maintainers who aren't actually the creators of this project are repeatedly refusing to make any changes.

People don't want to contribute on projects where maintaners don't even care. Start by creating a PR template, at least it will let people know that it's accepting contributions, make formal bug report template. If it's open-source atleast make it feel like one.

**Comment by vladimyr** [comment]
> I am pretty sure that even without all that exploitation, had I informed you all about this, no one could have cared any less about this issue.

Good for you, keep growing your prejudice. You know, you could avoid **guessing** by just doing it. 

> Maybe it would have been more polite, but, nothing more.

FYI keeping things polite is important enough on its own!

> Now that everyone has got a taste of Spam, they know the problem.

Real eye-opener! Let's ignore the fact it is known and discussed multiple times before https://github.com/echojs/echojs/issues/1#issue-407634271 But yeah, you've done something brand new here. In contrast to everyone else who contributed with their helpful and constructive thoughts, you decided to not follow the suit and instead _contributed_ with abuse.

> Maintainers who aren't actually the creators of this project are repeatedly refusing to make any changes.

Would you be so kind to back up this claim? I failed to find those on-site/Github discussions that might prove it. Or is it just your style to throw in a bunch of judgment and accusations?
Also, I don't remember when did _not acting_ on something (due to various possible reasons) become equal to refusing?

> People don't want to contribute on projects where maintaners don't even care.

Do you call this contribution? And I'm not talking about raising an issue, I'm talking about the whole charade that precedes it.

> Start by creating a PR template, at least it will let people know that it's accepting contributions, make formal bug report template. If it's open-source atleast make it feel like one.

So we are back to demanding things again. This is not a proposal or conversation starter, this is direct demand towards maintainers aka _let me tell you what to do_.

**Comment by theabbie** [comment]
@vladimyr 
> Would you be so kind to back up this claim? I failed to find those on-site/Github discussions that might prove it. Or is it just your style to throw in a bunch of judgment and accusations?

That's the beauty of git, you can see who has created what and it's obvious that @antirez has built most of it, after that, @fallanic is a maintainer but not the creator, @tracker1 is active here, but, still not a creator.

Before this, EchoJS had only been exposed to off-topic posts, mostly done manually, and, not necessarily by one person, This was actual Spam.

> So we are back to demanding things again. This is not a proposal or conversation starter, this is direct demand towards maintainers aka let me tell you what to do.

Whether you call it a demand, or suggestion, up to you. You know it's a good practice to have them, Statements starting with "should", like, "You should add CAPTCHA" is not considered demand, it's a suggestion, Not a request either.

Though no one is required to improve EchoJS, it will be for everyone's benefit. It has been an old Community, so, it's understandable why it doesn't have Spam protection, things have changed now.

**Comment by vladimyr** [comment]
> > Would you be so kind to back up this claim? I failed to find those on-site/Github discussions that might prove it. Or is it just your style to throw in a bunch of judgment and accusations?
> 
> That's the beauty of git, you can see who has created what and it's obvious that @antirez has built most of it, after that, @fallanic is a maintainer but not the creator, @tracker1 is active here, but, still not a creator.

Indeed you can, one might start with this PR https://github.com/echojs/echojs/pull/4
Your response does not present any discussions that imply **refusal** of implementing antispam mechanisms! Again, not implementing does not equal (actively) refusing!

(There goes my hope you wouldn't waste Salvatore's time by dragging him into this mess...)

> Before this, EchoJS had only been exposed to off-topic posts, mostly done manually, and, not necessarily by one person, This was actual Spam.

Did you just assume folks weren't able to figure out that it could be automated until you came and showed us all it can be done?! :open_mouth: 

> Whether you call it a demand, or suggestion, up to you. You know it's a good practice to have them, Statements starting with "should", like, "You should add CAPTCHA" is not considered demand, it's a suggestion, Not a request either.

I'm trying really hard to find those _should/s_ inside your statement:

> Start by creating a PR template, at least it will let people know that it's accepting contributions, make formal bug report template. If it's open-source atleast make it feel like one.

Could you help me with that?

> Though no one is required to improve EchoJS, it will be for everyone's benefit. It has been an old Community, so, it's understandable why it doesn't have Spam protection, things have changed now.

I agree with the premise but malice and arrogance aren't good tools for bringing change. One should know better and because of that and because I have more important things to do than keep this pointless conversation going I'm checking out. Enjoy your mess, hope you'll gonna learn something out of it...

**Comment by theabbie** [comment]
@vladimyr Sure, this thread has been unnecessarily long and won't lead to anything, Hoping EchoJS becomes better in future, or, some angel creates a PR with everything fixed. Have a good day.

**Comment by fallanic** [comment]
Hey @theabbie 

You're making good points, we should add more validation mechanism to prevent spammers from signing up, it's too easy to create accounts for abusers right now.

@tracker1 and @vladimyr are also making great points, this is an open source repository, community managed website, both ideas and PRs are welcome. 

Like theabbie said, I'm just a maintainer, working on this on my free time, and hosting the website on my own dime. 
I added a few anti-spam tools at some point because they're were easy fixes and did the job until now, but they won't work if a spammer is insistant and/or resourceful.

Closing this thread, in favor of another issue with some of your suggestions @theabbie
